### PR TITLE
Feat password reset.

### PR DESCRIPTION
**Pull request by clemensmanert**
- Http api for a request to reset the password
- An email template which is sent on a password reset request
- Some templates where the user can actually reset the password

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -9,10 +9,15 @@ class RunConfig():
     CLIENT_NAME = 'BuzznClient'
     SQLALCHEMY_TRACK_MODIFICATIONS = False
     SQLALCHEMY_DATABASE_URI = environ.get('BUZZN_SQLALCHEMY_DATABASE_URI')
-    PASSWORD_SALT = environ.get('PASSWORD_SALT')
+    PASSWORD_SALT = environ.get('BUZZN_PASSWORD_SALT')
```
Comment (missgreenwood) [inline app.py:12]: Please document this also in the README.

File: app.py [hunk 2]
```diff
@@ -9,10 +9,15 @@ class RunConfig():
...
     JWT_SECRET_KEY = environ.get('JWT_SECRET_KEY')
     OAUTH2_REFRESH_TOKEN_GENERATOR = True
-
+    BUZZN_SMTP_SERVER = environ.get('BUZZN_SMTP_SERVER')
```
Comment (missgreenwood) [inline app.py:15]: Please document all of these new environment variables in the README.

File: models/user.py [hunk 3]
```diff
@@ -1,9 +1,21 @@
...
 from enum import Enum
+import secrets
+
 import bcrypt
-from flask import current_app
+from flask import current_app as app
 from sqlalchemy import ForeignKey
 from util.database import db
 
+PASSWORD_MAX_LENGTH = 7
```
Comment (missgreenwood) [inline models/user.py]: Didn't we agree on 8?

File: models/user.py [hunk 4]
```diff
@@ -1,9 +1,21 @@
...
+
+
+class GenderType(Enum):
```
Comment (missgreenwood) [inline models/user.py]: How is this relevant for our app? Why not leave it out/neutral?

File: models/user.py [hunk 5]
```diff
@@ -41,44 +55,70 @@ def generate_password_hash(target):
...
+        """Checks whether the given password matches the password length
+            condition.
+        :param str new_password: password to check in plaintext
+        :returns: True if the given password is too long, False otherwise.
+        :rtype: bool
+        """
+        return len(new_password) > PASSWORD_MAX_LENGTH
 
     _id = db.Column(db.Integer, primary_key=True)
+    _gender = db.Column(db.Enum(GenderType))
```
Comment (missgreenwood) [inline models/user.py]: See above. How is gender relelvant?

File: models/user.py [hunk 6]
```diff
@@ -41,44 +55,70 @@ def generate_password_hash(target):
...
     _password = db.Column(db.String(333))
-    _status = db.Column(db.Enum(ActiveType))
+    _state = db.Column(db.Enum(StateType))
     _role = db.Column(db.Enum(RoleType))
     _meter_id = db.Column(db.String(32))
     _inhabitants = db.Column(db.Integer)
     _flat_size = db.Column(db.Float)
     _group_id = db.Column(db.Integer, ForeignKey('group._id'))
+    _password_reset_token = db.Column(db.String(33), unique=True)
+    _password_reset_token_expiries = db.Column(db.DateTime)
```
Comment (missgreenwood) [inline models/user.py]: Grammatically correct would be "_password_reset_token_expires".

File: models/user.py [hunk 7]
```diff
@@ -41,44 +55,70 @@ def generate_password_hash(target):
...
 
-    def __init__(self, name, activation_token, meter_id, group_id):
+    # Plain value constructor, too many arguments is ok here
+    #pylint: disable=too-many-arguments
+    def __init__(self, gender, name, mail, activation_token, meter_id, group_id):
```
Comment (missgreenwood) [inline models/user.py]: Cf. what I said before about gender.

File: models/user.py [hunk 8]
```diff
@@ -41,44 +55,70 @@ def generate_password_hash(target):
...
         :param str meter_id: the user's meter id
-        :parem int group_id: the user's group id
+        :param int group_id: the user's group id
         """
+        self._gender = gender
         self._name = name
+        self._mail = mail
         self._activation_token = activation_token
-        self._status = ActiveType.ACTIVATION_PENDING
+        self._state = StateType.ACTIVATION_PENDING
```
Comment (missgreenwood) [inline models/user.py]: Very good, changing the state type name!

File: models/user.py [hunk 9]
```diff
@@ -41,44 +55,70 @@ def generate_password_hash(target):
...
         self._role = RoleType.LOCAL_POWER_TAKER
         self._meter_id = meter_id
         self._group_id = group_id
 
     def get_id(self):
         return self._id
 
+    def get_gender(self):
```
Comment (missgreenwood) [inline models/user.py]: Cf. gender.

File: models/user.py [hunk 10]
```diff
@@ -13,14 +25,16 @@ class RoleType(Enum):
...
+class StateType(Enum):
     """Indicates the user account's state in its lifecycle.
     """
     ACTIVATION_PENDING = 'User activation pending.'
+    PASSWORT_RESET_PENDING = 'User requested a password change'
     ACTIVE = 'User account is active.'
     DISABLED = 'User account is disabled'
 
-
+# Maybe setters should not count as public.
```
Comment (missgreenwood) [inline models/user.py:36]: Getter/setter methods are not really Pythonic. Consider making some attributes public to access them easily with the "."-notation.

File: models/user.py [hunk 11]
```diff
@@ -124,7 +170,25 @@ def set_password(self, newPassword):
...
+    def set_state(self, new_state):
+        self._state = new_state
+
+    def generate_password_request_token(self):
+        """Generats a new password reset token and sets the password token
+        expiry date.
+        :returns: The created token.
+        :rtype: str"""
+        self._password_reset_token = secrets.token_hex(33)
+        self._password_reset_token_expiries = datetime.now() + timedelta(days=1)
```
Comment (missgreenwood) [inline models/user.py]: "_password_reset_token_expires"

File: util/mailer.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import smtplib
+import ssl
+
+from flask import current_app as app
+
+
+def send_mail(receiver, message):
+    """Sends a new mail to the given receiver.
+    :param receiver: The receiver's mail address.
```
Comment (missgreenwood) [inline util/mailer.py]: Please use the same syntax as before: ":param str receiver: The receiver's mail address."

File: util/mailer.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    :type receiver: str
+    :param message: The message.
```
Comment (missgreenwood) [inline util/mailer.py]: Same here.

File: util/mailer.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        context = ssl.create_default_context()
+
+        with smtplib.SMTP_SSL(app.config['BUZZN_SMTP_SERVER'],
+                              app.config['BUZZN_SMTP_SERVER_PORT'],
+                              context=context) as server:
+            server.login(app.config['BUZZN_EMAIL'],
+                         app.config['BUZZN_EMAIL_PASSWORD'])
+        server.sendmail(app.config['BUZZN_EMAIL'], receiver, message)
+    else:
+        raise AssertionError("BUZZN_MAILER not set, no mailer configured. Can not send mail.")
```
Comment (missgreenwood) [inline util/mailer.py]: "Cannot"

File: templates/password/layout.html [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  {% block head %}
+  {% endblock %}
+</head>
+
+<body>
+  {% block body %}
+  {% endblock %}
+</body>
+
+</html>
```
Comment (missgreenwood) [inline templates/password/layout.html]: There are some files without newlines at the end in this PR. Please consider changing this.

File: templates/password/reset_password_mail.txt [hunk 16]
```diff
@@ -0,0 +1,27 @@
+Subject: Passwort zurücksetzen für Buzzn-App
+
+{{greeting}},
+wir haben eine Anfrage zum Zurücksetzen des Passworts erhalten.
```
Comment (missgreenwood) [inline templates/password/reset_password_mail.txt:4]: "Wir"

File: templates/password/reset_password_mail.txt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+Folgen Sie diesem Link um ein neues Passwort anzugeben:
+
+{{link}}
+
+Sollten Sie kein neues Passwort angefordert haben, können Sie diese diese
```
Comment (missgreenwood) [inline templates/password/reset_password_mail.txt]: "diese" ist doppelt

File: templates/password/reset_password_mail.txt [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+Mail ignorieren.
+
+Mit freundlichen Grüßen
+Ihr Buzzn Team
+
+--
+
+BUZZN – People Power
+Tel. 089-4161714-32
+Fax.089-4161714-88
```
Comment (missgreenwood) [inline templates/password/reset_password_mail.txt]: "Fax 089[...]"

File: tests/buzzn_test_case.py [hunk 19]
```diff
@@ -13,12 +13,18 @@ class TestConfig():
...
     SQLALCHEMY_TRACK_MODIFICATIONS = False
     SQLALCHEMY_DATABASE_URI = 'sqlite:///:memory:'
-    PASSWORD_SALT = bcrypt.gensalt()
-
+    PASSWORD_SALT = bcrypt.gensalt().hex()
+    BUZZN_SMTP_SERVER = 'BUZZN_SMTP_SERVER'
+    BUZZN_SMTP_SERVER_PORT = '456'
+    BUZZN_EMAIL = 'team@buzzn.net'
+    BUZZN_EMAIL_PASSWORD = 'BUZZN_EMAIL_PASSWORD'
+    BUZZN_BASE_URL = 'localhost:5000'
```
Comment (missgreenwood) [inline tests/buzzn_test_case.py:21]: What are the effects of this line?

File: tests/test_set_password.py [hunk 20]
```diff
@@ -43,8 +42,8 @@ def test_token_does_not_match(self):
 
     def test_no_activation_pending(self):
         """Expect an error if an account is activated for the seconds time."""
-        db.session.add(User("SomeUser", "SomeToken",
-                            "SomeMeterId", "SomeGroup"))
+        db.session.add(User(GenderType.MALE, "SomeUser", "User@Some.net",
```
Comment (missgreenwood) [inline tests/test_set_password.py]: Do you really want to create a user with the same parameters as in "test_user_does_not_exist()"? I'd suggest creating all user instances needed for the different test functions in a setUp() method to avoid duplicate code.

File: routes/reset_password.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+from util.error import Error
+from util.translation import get_opening_greeting
+from util.mailer import send_mail
+
+ResetPassword = Blueprint('ResetPassword', __name__)
+
+
+class Errors:
+    UNKNOWN_USER = Error('Unknown user', 'No user found')
+    DISABLED_USER = Error('User account disabled',
```
Comment (missgreenwood) [inline routes/reset_password.py]: Better: "deactivated".

File: routes/reset_password.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+def reset_password(token):
+    return Response(render_template('password/request.html',
+                                    passwordResetToken=token))
+
+
+@ResetPassword.route('/password/reset/<token>', methods=['POST'])
+def do_password(token):
+    """Resets the password of the user account according to the given token.
+
+    :param token: The account with this token will set the password.
```
Comment (missgreenwood) [inline routes/reset_password.py]: Please provide the param types.

File: routes/reset_password.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+class Errors:
+    UNKNOWN_USER = Error('Unknown user', 'No user found')
+    DISABLED_USER = Error('User account disabled',
+                          'Can not request new password for disabled account.')
+
+
+@ResetPassword.route('/password/request-reset-token', methods=['POST'])
+def request_password_reset_token():
+    """Requests a new password token. Sends a mail to the user's mail address.
+    :param user: The user's mail address.
```
Comment (missgreenwood) [inline routes/reset_password.py]: Please provide the param type.

File: tests/buzzn_test_case.py (hunk 19 above)
Comment (clemensmanert) [reply tests/buzzn_test_case.py:21]: `Baseurl of the backend i.e. address. Everyhing which needs to be in front of the api calls. Example mybuzzn-backend.buzzn.net` I also added this to the readme :)
